src/solvers/petsc_dm_wrapper.C

```diff
@@ -57,40 +373,273 @@ void PetscDMWrapper::init_and_attach_petscdm(System & system, SNES & snes)
 
   PetscErrorCode ierr;
 
-  // Eventually, we'll traverse the mesh hierarchy and cache
-  // for each grid level, but for now, we're just using the
-  // finest level
-  unsigned int n_levels = 1;
-  this->init_dm_data(n_levels);
+  MeshBase & mesh = system.get_mesh();   // Convenience
+  MeshRefinement mesh_refinement(mesh); // Used for swapping between grids
+
+  // Theres no need for these code paths while traversing the hierarchy
+  mesh.allow_renumbering(false);
+  mesh.allow_remote_element_removal(false);
+  mesh.partitioner() = nullptr;
+
+  // First walk over the active local elements and see how many maximum MG levels we can construct
+  unsigned int n_levels = 0;
+  for ( auto & elem : mesh.active_local_element_ptr_range() )
+    {
+      if ( elem->level() > n_levels )
+        n_levels = elem->level();
+    }
+  // On coarse grids some processors may have no active local elements,
+  // these processors shouldnt make projections
+  if (n_levels >= 1)
+    n_levels += 1;
+
+  // How many MG levels did the user request?
+  int usr_requested_mg_lvls = 0;
+  usr_requested_mg_lvls = command_line_next("-pc_mg_levels", usr_requested_mg_lvls);
+
+  // Only construct however many levels were requested if something was actually requested
+  if ( usr_requested_mg_lvls != 0 )
+    {
+      // Dont request more than avail num levels on mesh, require at least 2 levels
+      libmesh_assert_less_equal( (unsigned int)usr_requested_mg_lvls, n_levels );
```

Comment:
C++ style `static_cast<unsigned int>` here.

Reply:
Or could we just declare that as unsigned int to begin with?